### PR TITLE
Implement support for github.com/jackc/pgx

### PR DESCRIPTION
**Pull request by hallabro**
These changes add support for `github.com/jackc/pgx` (including `pgxpool`).

Two of the biggest changes with this implementation:
1. The `pgx` functions return interface types. Because of this we need to check for `instr.Call.Method` calls as well `instr.Call.Value`.
2. For the same reason we need to check for `Named` types in addition to `Pointer` types.

There was an issue with the tests in `pkg/analyzer` that would not allow the tests to run with external packages. From the documentation for `analysistest.Run`:

> // If the directory contains a go.mod file, Run treats it as the root of the
> // Go module in which to work. Otherwise, Run treats it as the root of a
> // GOPATH-style tree, with package contained in the src subdirectory.

The solution was to add a `go.mod` and vendor in the `pkg/analyzer/testdata` directory. This leads to some packages being vendored twice unfortunately but I see no easy solution around this.

This is the first time I write some linter/static code analyzer tool so any feedback is appreciated.

Thanks.

Closes https://github.com/ryanrolds/sqlclosecheck/issues/23.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,12 +21,3 @@ go vet -vettool=$(which sqlclosecheck) ./...
...
 ```
-
-## Roadmap
-
-* ~~Get linter working~~
-* ~~Added some basic test cases~~
-* ~~Require that Close be deferred~~
-* ~~Add sqlx checking~~
-* ~~Test across a bunch of projects~~
-* ~~Introduce linter to [golangci-lint](https://github.com/golangci/golangci-lint).~~
```
Comment (hallabro) [inline README.md:32]: Just some cleanup while I was here. I feel that a finished roadmap doesn't add much value to the README but let me know if you want to keep it.

**Comment by ryanrolds** [comment]
Thank you. I will take a look at this by the end of the weekend.

File: pkg/analyzer/analyzer.go [hunk 2]
```diff
@@ -88,17 +87,22 @@ func run(pass *analysis.Pass) (interface{}, error) {
...
-func getTargetTypes(pssa *buildssa.SSA, targetPackages []string) []*types.Pointer {
-	targets := []*types.Pointer{}
+func getTargetTypes(pssa *buildssa.SSA, targetPackages []string) []any {
+	targets := []any{}
 
 	for _, sqlPkg := range targetPackages {
 		pkg := pssa.Pkg.Prog.ImportedPackage(sqlPkg)
 		if pkg == nil {
 			// the SQL package being checked isn't imported
-			return targets
```
Comment (ryanrolds) [inline pkg/analyzer/analyzer.go:98]: 🤔

File: pkg/analyzer/analyzer.go [hunk 3]
```diff
@@ -63,16 +64,14 @@ func run(pass *analysis.Pass) (interface{}, error) {
...
+				pointerTargetValues := getTargetTypesValues(b, i, targetTypes)
+				if len(pointerTargetValues) == 0 {
 					continue
 				}
 
-				// log.Printf("%s", f.Name())
-
 				// For each found target check if they are closed and deferred
-				for _, targetValue := range targetValues {
+				for _, targetValue := range pointerTargetValues {
```
Comment (ryanrolds) [inline pkg/analyzer/analyzer.go:74]: These are only sometimes pointers now, right? Should "pointer" still be added here? Am I missing something?

**Review by ryanrolds — CHANGES_REQUESTED** [review]
Thank you. This is overall great work, and look forward to merging it. 😸 

I have one question about a variable rename and would like to see some of the new code DRYed out a bit. Happy to chat about strategies to eliminate the repeated code. This is my first linter too,  I'm learning with each change as well.

File: pkg/analyzer/analyzer.go [hunk 4]
```diff
@@ -149,25 +169,51 @@ func getTargetTypesValues(b *ssa.BasicBlock, i int, targetTypes []*types.Pointer
...
+								instr: call,
+							})
+						}
+					case ssa.Value:
+						if types.Identical(instr.Type(), tt) {
+							targetValues = append(targetValues, targetValue{
+								value: &instr,
+								instr: call,
+							})
+						}
```
Comment (ryanrolds) [inline pkg/analyzer/analyzer.go:216]: Example of area with duplicated logic.

File: pkg/analyzer/analyzer.go [hunk 5]
```diff
@@ -265,9 +310,18 @@ func getAction(instr ssa.Instruction, targetTypes []*types.Pointer) action {
...
+				if types.Identical(instrType, tt) {
+					if checkClosed(instr.Referrers(), targetTypes) {
+						return actionHandled
+					}
+				}
+			case *types.Named:
+				if types.Identical(instrType, tt) {
+					if checkClosed(instr.Referrers(), targetTypes) {
+						return actionHandled
+					}
```
Comment (ryanrolds) [inline pkg/analyzer/analyzer.go:324]: Another example. There are a couple more areas. Don't worry about the `Value` and `Method` paths.

**Comment by hallabro** [comment]
Thanks for the feedback Ryan! I'll have a look at your comments today.

File: pkg/analyzer/analyzer.go (hunk 2 above)
Comment (hallabro) [reply pkg/analyzer/analyzer.go:98]: Identical to https://github.com/ryanrolds/sqlclosecheck/pull/24 but `continue` probably makes more sense. 🙂

I will rebase.

File: pkg/analyzer/analyzer.go (hunk 5 above)
Comment (hallabro) [reply pkg/analyzer/analyzer.go:324]: Can you please elaborate on the `Value` and `Method` paths? They don't seem to be part of the code that you highlighted in your comment.

**Comment by hallabro** [comment]
I have pushed some changes around the duplicate code. Thanks for pointing it out! All this type casting is making it a little bit difficult for me to fully understand what's going on sometimes. 

Hopefully I've responded to your other comments as well. Let me know what you think.

I'd happily continue to work on the code until it reaches a mergeable state.

**Comment by ryanrolds** [comment]
I was thinking something like:
```
func getTargetTypesValues(b *ssa.BasicBlock, i int, targetTypes []any) []targetValue {
	targetValues := []targetValue{}

	instr := b.Instrs[i]
	call, ok := instr.(*ssa.Call)
	if !ok {
		return targetValues
	}

	signature := call.Call.Signature()
	results := signature.Results()
	for i := 0; i < results.Len(); i++ {
		v := results.At(i)
		varType := v.Type()

		for _, targetType := range targetTypes {
			switch tt := targetType.(type) {
			case *types.Pointer:
				targetValues = checkAndAddTargetValue(varType, tt, call, targetValues)
			case *types.Named:
				targetValues = checkAndAddTargetValue(varType, tt, call, targetValues)
			}
		}
	}

	return targetValues
}

func checkAndAddTargetValue(varType types.Type, tt types.Type, call *ssa.Call, targetValues []targetValue) []targetValue {
	if !types.Identical(varType, tt) {
		return targetValues
	}

	for _, cRef := range *call.Referrers() {
		switch instr := cRef.(type) {
		case *ssa.Call:
			if len(instr.Call.Args) >= 1 && types.Identical(instr.Call.Args[0].Type(), tt) {
				targetValues = append(targetValues, targetValue{
					value: &instr.Call.Args[0],
					instr: call,
				})
			}
		case ssa.Value:
			if types.Identical(instr.Type(), tt) {
				targetValues = append(targetValues, targetValue{
					value: &instr,
					instr: call,
				})
			}
		}
	}

	return targetValues
}
```

**Review by ryanrolds — APPROVED** [review]

**Comment by ryanrolds** [comment]
Thank you. :) Great job.

**Pull request merged**